### PR TITLE
CI: Mdbook link checker

### PR DESCRIPTION
**Pull request by 4ydan**
#555 closed to change target branch. All links fixed, link checker can be integrated into CI.
The remaining links were actually way less effort than I thought, but the last link fix session scared me of finishing this ^^.
Looking forward to no more broken links inside the book.

Maybe open follow up issue to enable external link check.

Weird ci [error](https://build.libelektra.org/blue/organizations/jenkins/PermaPlanT-Folder%2FPermaplanT/detail/PR-604/6/pipeline/194)

## Basics

- [x] I added a line to [/doc/CHANGELOG.md](/doc/CHANGELOG.md)
- [x] The PR is rebased with current master.
- [x] Details of what you changed are in commit messages.
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildserver is happy.

## Checklist

- [x] I have installed and I am using [pre-commit hooks](../doc/contrib/README.md#Hooks)
- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added unit tests for my code
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
      (exceptions are documented)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Examples are well chosen and understandable

### CODE REVIEW

**Comment by 4ydan** [comment]
Jenkins build please

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -34,7 +34,7 @@ Otherwise please check these points when getting a PR done:
 - [ ] I added unit tests for my code
 - [ ] I added code comments, logging, and assertions as appropriate
 - [ ] I translated all strings visible to the user
-- [ ] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
+- [ ] I mentioned [every code or binary](https://github.com/ElektraInitiative/PermaplanT/blob/master/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
```
Comment (markus2330) [inline .github/PULL_REQUEST_TEMPLATE.md:37]: Hopefully we soon get git.permaplant.net redirection links but let us keep it like this for now.

File: book.toml [hunk 2]
```diff
@@ -13,3 +13,42 @@ command = "mdbook-mermaid"
 
 [output.html]
 additional-js = ["doc/mermaid.min.js", "doc/mermaid-init.js"]
+
+[output.linkcheck]
+# Should we check links on the internet? Enabling this option adds a
+# non-negligible performance impact
+follow-web-links = false
```
Comment (markus2330) [inline book.toml:20]: Still open: did you try it?

```suggestion
follow-web-links = true
```

File: book.toml [hunk 3]
```diff
@@ -13,3 +13,42 @@ command = "mdbook-mermaid"
...
+traverse-parent-directories = false
+
+# If necessary, you can exclude one or more links from being checked with a
+# list of regular expressions. The regex will be applied to the link href (i.e.
+# the `./index.html` in `[some page](./index.html)`) so it can be used to
+# ignore both web and filesystem links.
+#
+# Hint: you can use TOML's raw strings (single quote) to avoid needing to
+# escape things twice.
+exclude = [ 'google\.com' ]
```
Comment (markus2330) [inline book.toml:33]: Why would we need this exclusion?

```suggestion
exclude = [ ]
```

File: book.toml [hunk 4]
```diff
@@ -13,3 +13,42 @@ command = "mdbook-mermaid"
...
 additional-js = ["doc/mermaid.min.js", "doc/mermaid-init.js"]
+
+[output.linkcheck]
+# Should we check links on the internet? Enabling this option adds a
+# non-negligible performance impact
+follow-web-links = false
+
+# Are we allowed to link to files outside of the book's root directory? This
+# may help prevent linking to sensitive files (e.g. "../../../../etc/shadow")
+traverse-parent-directories = false
```
Comment (markus2330) [inline book.toml:24]: Wouldn't have this fixed the problems of links going outside of doc?

File: book.toml [hunk 5]
```diff
@@ -13,3 +13,42 @@ command = "mdbook-mermaid"
...
+# The number of seconds a cached result is valid for (12 hrs by default)
+cache-timeout = 43200
+
+# How should warnings be treated?
+#
+# - "warn" will emit warning messages
+# - "error" treats all warnings as errors, failing the linkcheck
+# - "ignore" will ignore warnings, suppressing diagnostic messages and allowing
+#   the linkcheck to continuing
+warning-policy = "ignore"
```
Comment (markus2330) [inline book.toml:47]: Still on ignore?

```suggestion
warning-policy = "error"
```

File: ci/container-images/permaplant-mdbook/Dockerfile [hunk 6]
```diff
@@ -1,5 +1,6 @@
 FROM rust:1.67.1-slim
 
-RUN apt update && apt install -y libpq-dev
-RUN cargo install mdbook mdbook-mermaid
+RUN apt update && apt install -y libpq-dev curl pkg-config ca-certificates build-essential make perl gcc libc6-dev
+RUN cargo install mdbook mdbook-mermaid --verbose
```
Comment (markus2330) [inline ci/container-images/permaplant-mdbook/Dockerfile:4]: Is this needed for devs?

File: doc/CHANGELOG.md [hunk 7]
```diff
@@ -14,7 +14,7 @@ Syntax: `- short text describing the change _(Your Name)_`
 - Fix wrong placement of plantings _(Paul)_
 - _()_
 - CI: Send email when master fails #109 _(4ydan)_
-- _()_
+- CI: Add link checker to CI pipeline #555 _(4ydan)_
```
Comment (markus2330) [inline doc/CHANGELOG.md:17]: I didn't see where the check in CI runs, does this PR already add the link check to the CI pipeline?

**Review by markus2330 — APPROVED** [review]
Great job, cannot await to get this merged. Probably some settings need to be changed to make the CI link check effective? Is this planned as follow-up?

**Comment by 4ydan** [comment]
The link check is happening with mdbook build, a follow up issue was created for enabling external link checks.

File: book.toml (hunk 2 above)
Comment (4ydan) [reply book.toml:20]: Yes, too many errors, I opened a follow up issue to enable this.

File: book.toml (hunk 3 above)
Comment (4ydan) [reply book.toml:33]: It was a default setting.

File: book.toml (hunk 4 above)
Comment (4ydan) [reply book.toml:24]: Hm, will have to look into this another time, for now I replaced outside links with github links.

File: book.toml (hunk 5 above)
Comment (4ydan) [reply book.toml:47]: There is too many warnings to enable it on error. Every bracket used in any .md file without a link afterwards is a warning.

File: ci/container-images/permaplant-mdbook/Dockerfile (hunk 6 above)
Comment (4ydan) [reply ci/container-images/permaplant-mdbook/Dockerfile:4]: curl pkg-config ca-certificates build-essential make perl gcc libc6-dev This was needed to perform the linkchecking. The --verbose not.

File: doc/CHANGELOG.md (hunk 7 above)
Comment (4ydan) [reply doc/CHANGELOG.md:17]: Yes, here.
image
Inside test-and-build-mdbook during mdbook build.

Comment (markus2330) [reply doc/CHANGELOG.md:17]: Great

**Comment by 4ydan** [comment]
I think this is mergable for now and we just open up two issues. One for external link checks (which already exists) and one to eventually reach "warning-policy: error", which I will go open now.

**Comment by markus2330** [comment]
Thx! :heart: 

> Ok I will use please merge to communicate when pr's are ready.
> I thought please merge is maybe a more urgent one.

You understood correctly but in your case nearly everything is somewhat "urgent" because as long as build checks are missing things might go wrong on master.

**Pull request merged**